### PR TITLE
Improve signal loop

### PR DESCRIPTION
**Pull request by viktorprogger**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Fixed issues  | ❌

### CODE REVIEW

File: src/Cli/SignalLoop.php [hunk 1]
```diff
@@ -44,9 +43,29 @@ class SignalLoop implements LoopInterface
...
+     * @param EventDispatcherInterface $dispatcher
+     * @param int $memorySoftLimit Soft RAM limit in bytes. The loop won't let you continue to execute the program if soft limit is reached. Zero means no limit.
+     * @param int[] $exitSignals pcntl signal codes to exit the loop
+     * @param int[] $suspendSignals pcntl signal codes to pause loop execution
+     * @param int[] $resumeSignals pcntl signal codes to resume loop execution
+     */
+    public function __construct(
+        EventDispatcherInterface $dispatcher,
+        $memorySoftLimit = 0,
+        $exitSignals = [1, 2, 15], // SIGHUP, SIGINT, SIGTERM
```
Comment (romkatsu) [inline src/Cli/SignalLoop.php:59]: is it better to use constants?

Comment (viktorprogger) [reply src/Cli/SignalLoop.php:59]: Constants are not available when pcntl module is not loaded. Pcntl module is optional for this package.

**Comment by viktorprogger** [comment]
WDYT, should we use here `LoggerInterface` to warn about reaching the limit, or leave just log warning without event dispatching?

File: src/Cli/SignalLoop.php [hunk 2]
```diff
@@ -44,9 +43,29 @@ class SignalLoop implements LoopInterface
...
+    /**
+     * @param EventDispatcherInterface $dispatcher
+     * @param int $memorySoftLimit Soft RAM limit in bytes. The loop won't let you continue to execute the program if soft limit is reached. Zero means no limit.
+     * @param int[] $exitSignals pcntl signal codes to exit the loop
+     * @param int[] $suspendSignals pcntl signal codes to pause loop execution
+     * @param int[] $resumeSignals pcntl signal codes to resume loop execution
+     */
+    public function __construct(
+        EventDispatcherInterface $dispatcher,
+        $memorySoftLimit = 0,
```
Comment (xepozz) [inline src/Cli/SignalLoop.php:58]: Please add typehints

File: src/Event/MemoryLimitReached.php [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    private int $actual;
+
+    public function __construct(int $limit, int $actual)
+    {
+        $this->limit = $limit;
+        $this->actual = $actual;
+    }
+
+    /**
+     * @return int
```
Comment (xepozz) [inline src/Event/MemoryLimitReached.php:19]: PHPDoc is redundant

**Comment by xepozz** [comment]
> WDYT, should we use here `LoggerInterface` to warn about reaching the limit along with the added event, or even leave just log warning without event dispatching?

I think it may be developed with decorator pattent

File: src/Cli/SignalLoop.php [hunk 4]
```diff
@@ -44,9 +43,29 @@ class SignalLoop implements LoopInterface
...
+     * @param EventDispatcherInterface $dispatcher
+     * @param int $memorySoftLimit Soft RAM limit in bytes. The loop won't let you continue to execute the program if soft limit is reached. Zero means no limit.
+     * @param int[] $exitSignals pcntl signal codes to exit the loop
+     * @param int[] $suspendSignals pcntl signal codes to pause loop execution
+     * @param int[] $resumeSignals pcntl signal codes to resume loop execution
+     */
+    public function __construct(
+        EventDispatcherInterface $dispatcher,
+        int $memorySoftLimit = 0,
+        array $exitSignals = [1, 2, 15], // SIGHUP, SIGINT, SIGTERM
```
Comment (samdark) [inline src/Cli/SignalLoop.php:59]: Do we need to be able to customize these?

**Comment by samdark** [comment]
That should be documented. Good feature overall.

**Review by samdark — APPROVED** [review]

**Pull request merged**